File: Backend/src/main/java/com/fairsplit/service/UserService.java
package com.fairsplit.service;

import com.fairsplit.model.User;
import com.fairsplit.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.Date;

@Service
public class UserService {
    
    @Autowired
    private UserRepository userRepository;
    
    /**
     * Get all users from the database
     * @return List of all users
     */
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }
    
    /**
     * Find a user by their ID
     * @param id the user ID to search for
     * @return an Optional containing the user if found
     */
    public Optional<User> getUserById(Long id) {
        return userRepository.findById(id);
    }
    
    /**
     * Find a user by their username
     * @param username the username to search for
     * @return an Optional containing the user if found
     */
    public Optional<User> getUserByUsername(String username) {
        return userRepository.findByUsername(username);
    }
    
    /**
     * Get users filtered by group ID
     * @param groupId the group ID to filter users by
     * @return List of users belonging to the specified group
     */
    public List<User> getFilteredUsers(Long groupId) {
        if (groupId == null) {
            return getAllUsers();
        }
        // Find users who belong to the specified group
        return userRepository.findByGroups_Id(groupId);
    }

    /**
     * Create a new user
     * @param user the user to create
     * @return the saved user
     */
    public User createUser(User user) throws IllegalArgumentException {
        // Check if username already exists
        if (userRepository.existsByUsername(user.getUsername())) {
            throw new IllegalArgumentException("Username already exists");
        }
        // Set initial values if not provided
        if (user.getAmount() == null) {
            user.setAmount(0.0);
        }
        if (user.getCreatedAt() == null) {
            user.setCreatedAt(new Date());
        }

        user.setUpdatedAt(new Date());
        user.setActive(true);
        
        return userRepository.save(user);
    }
    /**
     * Update a user's balance by adding an amount to their current balance
     * @param id the ID of the user to update
     * @param newAmount the amount to update to the user's current balance
     * @return the updated user if found, otherwise empty Optional
     */
    public Optional<User> updateUserBalance(Long id, double newAmount){
        Optional<User> userOptional = userRepository.findById(id);
        if (userOptional.isPresent()) {
            User user = userOptional.get();
            user.setAmount(newAmount);
            user.setUpdatedAt(new Date());
            return Optional.of(userRepository.save(user));
        }
        return Optional.empty();
    }

    /**
     * Add an amount to a user's current balance
     * @param id the ID of the user to update
     * @param amountToAdd the amount to add to the user's current balance
     * @return the updated user if found, otherwise empty Optional
     */
    public Optional<User> addUserAmount(Long id, double amountToAdd) {
        Optional<User> userOptional = userRepository.findById(id);
        if (userOptional.isPresent()) {
            User user = userOptional.get();
            double currentAmount = user.getAmount();
            double newAmount = currentAmount + amountToAdd;
            user.setAmount(newAmount);
            user.setUpdatedAt(new Date());
            return Optional.of(userRepository.save(user));
        }
        return Optional.empty();
    }

    /**
     * Update an existing user
     * @param id the ID of the user to update
     * @param userDetails the updated user details
     * @return the updated user if found, otherwise empty Optional
     * @throws IllegalArgumentException if trying to update to a username that already exists
     */
    public Optional<User> updateUser(Long id, User userDetails) throws IllegalArgumentException {
        Optional<User> user = userRepository.findById(id);
        if (user.isPresent()) {
            User existingUser = user.get();
            // Update fields if provided
            if (userDetails.getUsername() != null && !userDetails.getUsername().equals(existingUser.getUsername())) {
                // Check if the new username already exists for another user
                if (userRepository.existsByUsername(userDetails.getUsername())) {
                    throw new IllegalArgumentException("Username already exists");
                }
                existingUser.setUsername(userDetails.getUsername());
            }
            
            if (userDetails.getPassword() != null) {
                existingUser.setPassword(userDetails.getPassword());
            }
            
            if (userDetails.getFullname() != null) {
                existingUser.setFullname(userDetails.getFullname());
            }
            
            if (userDetails.getAmount() != null) {
                existingUser.setAmount(userDetails.getAmount());
            }
            
            // Update timestamp
            existingUser.setUpdatedAt(new Date());
            
            return Optional.of(userRepository.save(existingUser));
        }
        
        return Optional.empty();
    }
    
    /**
     * Delete a user by their ID
     * @param id the ID of the user to delete
     * @return true if the user was deleted, false if not found
     */
    public boolean deleteUser(Long id) {
        if (userRepository.existsById(id)) {
            userRepository.deleteById(id);
            return true;
        }
        return false;
    }

    /**
     * Check if a username already exists
     * @param username the username to check
     * @return true if the username exists, false otherwise
     */
    public boolean existsByUsername(String username) {
        return userRepository.existsByUsername(username);
    }
    
}

File: Node-Backend/src/services/userService.ts
import { supabase } from '../config/supabase.js';
import { User, CreateUserRequest, UpdateUserRequest, UserTable, PublicUserInfo } from '../types/user.js';

export class UserService {
  private readonly tableName = 'users';

  /**
   * Get all users from the database
   */
  async getAllUsers(): Promise<User[]> {
    const { data, error } = await supabase
      .from(this.tableName)
      .select('*')
      .eq('is_active', true)
      .order('created_at', { ascending: false });

    if (error) {
      throw new Error(`Failed to fetch users: ${error.message}`);
    }

    return data || [];
  }

  /**
   * Find a user by their ID
   */
  async getUserById(id: string): Promise<User | null> {
    const { data, error } = await supabase
      .from(this.tableName)
      .select('*')
      .eq('id', id)
      .eq('is_active', true)
      .single();

    if (error) {
      if (error.code === 'PGRST116') {
        return null; // User not found
      }
      throw new Error(`Failed to fetch user: ${error.message}`);
    }

    return data;
  }

  /**
   * Find a user by their username
   */
  async getUserByUsername(username: string): Promise<User | null> {
    const { data, error } = await supabase
      .from(this.tableName)
      .select('*')
      .eq('username', username)
      .eq('is_active', true)
      .single();

    if (error) {
      if (error.code === 'PGRST116') {
        return null; // User not found
      }
      throw new Error(`Failed to fetch user by username: ${error.message}`);
    }

    return data;
  }

  /**
   * Get public user information by username (safe for public access)
   */
  async getPublicUserByUsername(username: string): Promise<PublicUserInfo | null> {
    const { data, error } = await supabase
      .from(this.tableName)
      .select('id, username, fullname, is_active')
      .eq('username', username)
      .eq('is_active', true)
      .single();

    if (error) {
      if (error.code === 'PGRST116') {
        return null; // User not found
      }
      throw new Error(`Failed to fetch public user info: ${error.message}`);
    }

    return data;
  }

  /**
   * Get users filtered by group ID
   */
  async getFilteredUsers(groupId: string): Promise<User[]> {
    if (!groupId) {
      return this.getAllUsers();
    }

    const { data, error } = await supabase
      .from(this.tableName)
      .select(`
        *,
        group_user!inner(group_id)
      `)
      .eq('group_user.group_id', groupId)
      .eq('is_active', true)
      .order('created_at', { ascending: false });

    if (error) {
      throw new Error(`Failed to fetch users by group: ${error.message}`);
    }

    return data || [];
  }

  /**
   * Create a new user (assumes Supabase Auth user already exists)
   */
  async createUser(userData: CreateUserRequest, supabaseUserId: string): Promise<User> {
    // Check if username already exists
    const existingUser = await this.getUserByUsername(userData.username);
    if (existingUser) {
      throw new Error('Username already exists');
    }

    const now = new Date().toISOString();
    const newUser: Omit<UserTable, 'id'> = {
      username: userData.username,
      fullname: userData.fullname || null,
      amount: userData.amount || 0,
      is_active: true,
      created_at: now,
      updated_at: now
    };

    const { data, error } = await supabase
      .from(this.tableName)
      .insert([{ ...newUser, id: supabaseUserId }])
      .select()
      .single();

    if (error) {
      throw new Error(`Failed to create user: ${error.message}`);
    }

    return data;
  }

  /**
   * Update a user's balance to a specific amount
   */
  async updateUserBalance(id: string, newAmount: number): Promise<User | null> {
    const { data, error } = await supabase
      .from(this.tableName)
      .update({
        amount: newAmount,
        updated_at: new Date().toISOString()
      })
      .eq('id', id)
      .eq('is_active', true)
      .select()
      .single();

    if (error) {
      if (error.code === 'PGRST116') {
        return null; // User not found
      }
      throw new Error(`Failed to update user balance: ${error.message}`);
    }

    return data;
  }

  /**
   * Add an amount to a user's current balance
   */
  async addUserAmount(id: string, amountToAdd: number): Promise<User | null> {
    // First get current user to calculate new balance
    const currentUser = await this.getUserById(id);
    if (!currentUser) {
      return null;
    }

    const newAmount = currentUser.amount + amountToAdd;
    return this.updateUserBalance(id, newAmount);
  }

  /**
   * Update an existing user
   */
  async updateUser(id: string, userDetails: UpdateUserRequest): Promise<User | null> {
    // Check if username already exists (if username is being updated)
    if (userDetails.username) {
      const existingUser = await this.getUserByUsername(userDetails.username);
      if (existingUser && existingUser.id !== id) {
        throw new Error('Username already exists');
      }
    }

    const updateData: Partial<UserTable> = {
      updated_at: new Date().toISOString()
    };

    if (userDetails.username !== undefined) {
      updateData.username = userDetails.username;
    }
    if (userDetails.fullname !== undefined) {
      updateData.fullname = userDetails.fullname;
    }
    if (userDetails.amount !== undefined) {
      updateData.amount = userDetails.amount;
    }

    const { data, error } = await supabase
      .from(this.tableName)
      .update(updateData)
      .eq('id', id)
      .eq('is_active', true)
      .select()
      .single();

    if (error) {
      if (error.code === 'PGRST116') {
        return null; // User not found
      }
      throw new Error(`Failed to update user: ${error.message}`);
    }

    return data;
  }

  /**
   * Delete a user by their ID (soft delete)
   */
  async deleteUser(id: string): Promise<boolean> {
    const { error } = await supabase
      .from(this.tableName)
      .update({ is_active: false, updated_at: new Date().toISOString() })
      .eq('id', id);

    if (error) {
      throw new Error(`Failed to delete user: ${error.message}`);
    }

    return true;
  }

  /**
   * Check if a username already exists
   */
  async existsByUsername(username: string): Promise<boolean> {
    const user = await this.getUserByUsername(username);
    return user !== null;
  }

  /**
   * Get user by Supabase Auth ID
   */
  async getUserBySupabaseId(supabaseUserId: string): Promise<User | null> {
    return this.getUserById(supabaseUserId);
  }
} 